docs/manual/scala/guide/cluster/DomainModelling.md

```diff
@@ -1,19 +1,202 @@
+  def applyCommand(cmd: ShoppingCartCommnad) =
+    cmd match {
+      case UpdateItem(_, qty, replyTo) if qty < 0 =>
+        Effect.reply(replyTo)(Rejected("Quantity must be greater than zero"))
+
+      // an item is delete by setting it's quantity to 00
+      case UpdateItem(productId, 0, replyTo) if !items.contains(productId) =>
+        Effect.reply(replyTo)(Rejected("Cannot delete item that is not already in cart"))
+
+      case UpdateItem(productId, quantity, replyTo) =>
+        Effect
+          .persist(ItemUpdated(productId, quantity))
+          .thenReply(replyTo) { updatedCart => // updated cart is state updated after applying ItemUpdated
+            Accepted
+          }
+
+      // check it out
+      case Checkout(replyTo) =>
+        Effect
+          .persist(CheckedOut)
+          .thenReply(replyTo){ updatedCart => // updated cart is state updated after applying CheckedOut
+            Accepted
+          }
+
+      case Get(replyTo) =>
+        Effect.reply(replyTo)(CartState(items, status = "open"))
+  }
+}
+
+case class CheckedOutShoppingCart(items: Map[String, Int]) extends ShoppingCart {
+  def applyCommand(cmd: ShoppingCartCommnad) =
+    cmd match {
+      // CheckedOut is a final state, not mutations allowed
+      case UpdateItem(_, _, replyTo) =>
+        Effect.reply(replyTo)(Rejected("Cannot update a checked-out cart"))
+      // CheckedOut is a final state, not mutations allowed
+      case Checkout(replyTo) =>
+        Effect.reply(replyTo)(Rejected("Cannot checkout a checked-out cart"))
+
+      // it is allowed to read it's state though
+      case Get(replyTo) =>
+        Effect
+        .reply(replyTo)(CartState(items, status = "checked-out"))
+    }
+
+}
+```
+
+Command handlers are the meat of the model. They encode the business rules of your model and act as a guardian of the model consistency. Any mutation must be validated by it. However, they don't apply the mutations. Instead, they express the mutations in the form of persisted events.
+
+Because an Aggregate is intended to model a consistency boundary, it's not recommended to validate commands using data that's not available in scope. Any decision should be solely based on the data passed in the commands and the state of the aggregate. Any external call should be considered a smell because it means that the aggregate is not in full control of the invariants it's supposed to be protecting.
+
+You may have notice that there two ways of sending back a reply. Using `Effect.reply` and `Effect.persist(...).thenReply`. The first one is available directly on `Effect` and should be used when you reply without persisting any event. In this case, you can use the available state in scope because it's guaranteed to not have changed. The second variant should be used when you have persisted one or more events. The updated state is then made available to you on the function used to define the reply.
```

Comment:
```suggestion
There two ways of sending back a reply: using `Effect.reply` and `Effect.persist(...).thenReply`. The first one is available directly on `Effect` and should be used when you reply without persisting any event. In this case, you can use the available state in scope because it's guaranteed to not have changed. The second variant should be used when you have persisted one or more events. The updated state is then made available to you on the function used to define the reply.
```